documentation/intermediate-ca.md

```diff
@@ -0,0 +1,199 @@
+# Intermediate CA support with Puppet Server
+
+## Configuration guide
+
+When Puppet Server or PE is installed it will autogenerate a number of certificates. Regenerating all of these certificates can be a time consuming process, but this problem can be side stepped by generating an intermediate CA certificate that reuses the existing key pair that Puppet Server/PE has already created.
+
+### Configure the Puppet agent `certificate_revocation` checking
+
+Intermediate CA support is incompatible with full chain revocation checking, as Puppet is unable to load multiple CRLs. When an intermediate CA certificate is in use Puppet must have the `certificate_revocation` setting either set to `false`, or `leaf`. By default Puppet will attempt to perform full chain checking and will fail when encountering a non-root CA certificate, so this setting **must** be reconfigured on all Puppet agents before continuing.
+
+    puppet config set --section main certificate_revocation leaf
+
+See [PUP-3788](https://tickets.puppetlabs.com/browse/PUP-3788) for more information on chain revocation checking.
+
+### Generate an intermediate CA certificate request
+
+Save the following ruby script to a file, `csrgen.rb`. Make it executable (with `chmod u+x csrgen.rb`), and then run `./csrgen.rb`.
+
+    #!/opt/puppetlabs/puppet/bin/ruby
+    require 'puppet'
+    require 'openssl'
+
+    class Puppet::Application::GenCACSR < Puppet::Application
+      run_mode :master
+
+      def main
+        root_cert = OpenSSL::X509::Certificate.new(File.read(Puppet[:cacert]))
+        ca_key = OpenSSL::PKey::RSA.new(File.read(Puppet[:cakey]))
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        puts csr.to_s
+      end
+    end
+    Puppet::Application::GenCACSR.new.run
+
+Use this CSR to generate the intermediate CA certificate that Puppet will be using.
```

Comment:
What is the expected result of running `csrgen.rb`? When I run it, there is no output and no files are created.

Reply:
Wait, there's not output? @gguillotte if you're in the office you should come show me; a CSR should be printed to stdout. I can change this script to both print the CSR to stdout and save it to a file.